design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,227 @@
+      responses:
+        200:
+          description: Multipart upload initiated
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/MultipartUpload"
+        # ... other error responses
+
+  /repositories/{repository}/branches/{branch}/staging/multipart/{uploadId}:
+      parameters:
+      - in: path
+        name: repository
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: branch
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: uploadId
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: path
+        description: relative to the branch
+        required: true
+        schema:
+          type: string
+    put:
+      tags:
+        - staging
+      operationId: completeMultipartUpload
+      summary: Complete a multipart upload
+      description: Completes a multipart upload by assembling the uploaded parts.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/CompleteMultipartUpload"
+      responses:
+        200:
+          description: Multipart upload completed
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ObjectStats"
+        # ... other error responses
+
+    delete:
+      tags:
+        - staging
+      operationId: abortMultipartUpload
+      summary: Abort a multipart upload
+      description: Aborts a multipart upload.
+      responses:
+        204:
+          description: Multipart upload aborted
+
+```
+
+
+**Schemas used by operations**
+
+```yaml
+components:
+  schemas:
+    MultipartUpload:
+      type: object
+      properties:
+        upload_id:
+          type: string
+        locations:
+          type: array
+          items:
+            $ref: "#/components/schemas/StagingLocation"
+      required:
+        - upload_id
+        
+    UploadPart:
+      type: object
+      properties:
+        part_number:
+          type: integer
+        etag:
+          type: string
+      required:
+        - part_number
+        - etag
+        
+    CompleteMultipartUpload:
+      type: object
+      properties:
+        parts:
+          type: array
+          description: "List of uploaded parts, should be ordered by ascending part number"
+          items:
+            $ref: "#/components/schemas/UploadPart"
+      required:
+        - parts
+```
+
+### Support and discover
+
+Presign support is a capability lakectl discover before switching to use presign for upload or download from lakeFS.
+The multipart upload support will be part of the storage capability add `multipart_upload_support` optional field that when set to `true` the user can perform multipart upload using the new API.
```

Comment:
Why do we **actually** need this extra param?
Can't we rely on the block adapter type for now?

Reply:
It means that lakectl as a client will need to check:
1. s3 block adapter
2. presgined support enabled
3. lakefs version is above v1.x that include this feature

Reply:
For 1,2 the client can get this information from the GetConfig API
Regarding no. 3 I think is not necessary, IMHO it's perfectly fine for the server to return 501 and for the client to handle it appropriately.
Adding this field will cause ambiguity, since it is a direct result of blockstore_type + pre_sign

Reply:
I took elad's suggestion and like a feature flag it represents the capability. if we like to support multipart through lakefs API we can.

The second issue related to try and fail does not help with client side code when you like to sync files in parallel while trying to understand one of you can use this feature. possible but not clean.
Also status code for unsupported operation in the logs vs clean 200 ok, I prefer the last.